File: python/stathat.py
from urllib.parse import urlencode
from urllib.request import Request, urlopen

class StatHat:

        def http_post(self, path, data):
                pdata = urlencode(data)
                pdata = pdata.encode('ascii')
                req = Request('http://api.stathat.com' + path, pdata)
                resp = urlopen(req)
                return resp.read()

        def post_value(self, user_key, stat_key, value, timestamp=None):
                args = {'key': stat_key, 'ukey': user_key, 'value': value}
                if timestamp is not None:
                        args['t'] = timestamp
                return self.http_post('/v', args)

        def post_count(self, user_key, stat_key, count, timestamp=None):
                args = {'key': stat_key, 'ukey': user_key, 'count': count}
                if timestamp is not None:
                        args['t'] = timestamp
                return self.http_post('/c', args)

        def ez_post_value(self, ezkey, stat_name, value, timestamp=None):
                args = {'ezkey': ezkey, 'stat': stat_name, 'value': value}
                if timestamp is not None:
                        args['t'] = timestamp
                return self.http_post('/ez', args)

        def ez_post_count(self, ezkey, stat_name, count, timestamp=None):
                args = {'ezkey': ezkey, 'stat': stat_name, 'count': count}
                if timestamp is not None:
                        args['t'] = timestamp
                return self.http_post('/ez', args)

File: java/StatHat.java
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.BufferedReader;
import java.net.URL;
import java.net.URLConnection;
import java.net.URLEncoder;

class StatHat {

        private static void httpPost(String path, String data) {
                try {
                        URL url = new URL(path);
                        URLConnection conn = url.openConnection();
                        conn.setDoOutput(true);
                        OutputStreamWriter wr = new OutputStreamWriter(conn.getOutputStream());
                        wr.write(data);
                        wr.flush();

                        BufferedReader rd = new BufferedReader(new InputStreamReader(conn.getInputStream()));
                        String line;
                        while ((line = rd.readLine()) != null) {
                                System.out.println(line);
                        }
                        wr.close();
                        rd.close();
                }
                catch (Exception e) {
                        System.err.println(e);
                }
        }

        public static void postValue(String userKey, String statKey, Double value) {
                try {
                        String data = URLEncoder.encode("ukey", "UTF-8") + "=" + URLEncoder.encode(userKey, "UTF-8");
                        data += "&" + URLEncoder.encode("key", "UTF-8") + "=" + URLEncoder.encode(statKey, "UTF-8");
                        data += "&" + URLEncoder.encode("value", "UTF-8") + "=" + URLEncoder.encode(value.toString(), "UTF-8");
                        httpPost("http://api.stathat.com/v", data);
                }
                catch (Exception e) {
                        System.err.println("postValue exception:  " + e);
                }
        }

        public static void postCount(String userKey, String statKey, Double count) {
                try {
                        String data = URLEncoder.encode("ukey", "UTF-8") + "=" + URLEncoder.encode(userKey, "UTF-8");
                        data += "&" + URLEncoder.encode("key", "UTF-8") + "=" + URLEncoder.encode(statKey, "UTF-8");
                        data += "&" + URLEncoder.encode("count", "UTF-8") + "=" + URLEncoder.encode(count.toString(), "UTF-8");
                        httpPost("http://api.stathat.com/c", data);
                }
                catch (Exception e) {
                        System.err.println("postCount exception:  " + e);
                }
        }

        public static void ezPostValue(String ezkey, String statName, Double value) {
                try {
                        String data = URLEncoder.encode("ezkey", "UTF-8") + "=" + URLEncoder.encode(ezkey, "UTF-8");
                        data += "&" + URLEncoder.encode("stat", "UTF-8") + "=" + URLEncoder.encode(statName, "UTF-8");
                        data += "&" + URLEncoder.encode("value", "UTF-8") + "=" + URLEncoder.encode(value.toString(), "UTF-8");
                        httpPost("http://api.stathat.com/ez", data);
                }
                catch (Exception e) {
                        System.err.println("ezPostValue exception:  " + e);
                }
        }

        public static void ezPostCount(String ezkey, String statName, Double count) {
                try {
                        String data = URLEncoder.encode("ezkey", "UTF-8") + "=" + URLEncoder.encode(ezkey, "UTF-8");
                        data += "&" + URLEncoder.encode("stat", "UTF-8") + "=" + URLEncoder.encode(statName, "UTF-8");
                        data += "&" + URLEncoder.encode("count", "UTF-8") + "=" + URLEncoder.encode(count.toString(), "UTF-8");
                        httpPost("http://api.stathat.com/ez", data);
                }
                catch (Exception e) {
                        System.err.println("ezPostCount exception:  " + e);
                }
        }
}
